quantum/rgb_matrix/rgb_matrix_indicators.c

```diff
@@ -0,0 +1,122 @@
+    if (!rgb_matrix_indicators_kb()) return;
+    led_t led_state = host_keyboard_led_state();
+    for (size_t i = 0; i < rgb_matrix_indicators_config.matcher_count; i++) {
+        t_rgb_indicator_matcher matcher = rgb_matrix_indicators_config.matchers[i];
+        if (!(matcher.led_index >= 0 && matcher.led_index < RGB_MATRIX_LED_COUNT)) {
+            dprintf("LED %d is out of range of the rgb matrix configuration\n", matcher.led_index);
+        } else if (!(g_led_config.flags[matcher.led_index] & LED_FLAG_INDICATOR)) {
+            dprintf("LED %d is not configured as an indicator but has indicator config provided\n", matcher.led_index);
+        } else if ((matcher.led_state.raw & led_state.raw) == matcher.led_state.raw) {
+            uint8_t clamped_v = matcher.color.v;
+            if (!matcher.override_brightness_limit && matcher.color.v > rgb_matrix_config.hsv.v) clamped_v = rgb_matrix_config.hsv.v;
+            HSV color_hsv = {.h = matcher.color.h, .s = matcher.color.s, .v = clamped_v};
+            RGB color     = rgb_matrix_hsv_to_rgb(color_hsv);
+            rgb_matrix_set_color(matcher.led_index, color.r, color.g, color.b);
+        }
+    }
+}
+
+__attribute__((weak)) bool rgb_matrix_indicators_kb(void) {
+    return rgb_matrix_indicators_user();
+}
+
+__attribute__((weak)) bool rgb_matrix_indicators_user(void) {
+    return true;
+}
+
+void rgb_matrix_indicators_advanced(effect_params_t *params) {
+    /* special handling is needed for "params->iter", since it's already been incremented.
+     * Could move the invocations to rgb_task_render, but then it's missing a few checks
+     * and not sure which would be better. Otherwise, this should be called from
+     * rgb_task_render, right before the iter++ line.
+     */
+    RGB_MATRIX_USE_LIMITS_ITER(min, max, params->iter - 1);
+    rgb_matrix_indicators_advanced_kb(min, max);
+}
+
+__attribute__((weak)) bool rgb_matrix_indicators_advanced_kb(uint8_t led_min, uint8_t led_max) {
+    return rgb_matrix_indicators_advanced_user(led_min, led_max);
+}
+
+__attribute__((weak)) bool rgb_matrix_indicators_advanced_user(uint8_t led_min, uint8_t led_max) {
+    return true;
+}
+
+int add_rgb_matcher(t_rgb_indicator_matcher matcher) {
+    if (rgb_matrix_indicators_config.matcher_count >= RGB_INDICATOR_MATCHERS_COUNT_MAX) return -2;
+
+    uint8_t new_index = rgb_matrix_indicators_config.matcher_count;
+
+    rgb_matrix_indicators_config.matchers[new_index] = matcher;
+    rgb_matrix_indicators_config.matcher_count++;
+
+    eeconfig_flush_rgb_matrix_indicators(true);
+    return new_index;
+}
+
+int remove_rgb_matcher(uint8_t index) {
+    if (index >= rgb_matrix_indicators_config.matcher_count) return -1;
+
+    uint8_t right_matchers_count = (rgb_matrix_indicators_config.matcher_count - 1) - index;
+
+    memcpy(rgb_matrix_indicators_config.matchers + index, rgb_matrix_indicators_config.matchers + index + 1, sizeof(t_rgb_indicator_matcher) * right_matchers_count);
```

Comment:
Memcpy has undefined behaviour if SRC and DST overlap. Use memmove instead.

Reply:
Good point, thankyou! :heart: